---
title: "Basic population genomic and phylogenomic analyses"
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file,
 encoding=encoding,
 output_file=file.path(dirname(input_file), out_dir, 'index.html'))})
author: "Todd W. Pierson"
date: "23 October 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

In this tutorial...

Load the packages you need.
```{r, echo = TRUE, results = "hide", warning = FALSE, message = FALSE}
library(phangorn)
```

Read in one of the output files from the `ipyrad` assembly.
```{r}
dat <- read.phyDat("data/ubrucei.u.snps.phy", format="phy", type="DNA")
```

Create a distance matrix.
```{r}
dm.dat <- dist.ml(dat, exclude='pairwise')
```

Create a neighbor-joining tree for a quick peek at the data.
```{r}
treeNJ.dat <- NJ(dm.dat)
```

```{r}
# plot tree as unrooted
plot(treeNJ.dat, cex=, type='u')
```

